codec/src/main/java/io/netty/handler/codec/compression/AbstractClosableCompressionDecoder.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+/**
+ * A skeletal {@link ClosableCompressionDecoder} implementation.
+ */
+public abstract class AbstractClosableCompressionDecoder extends AbstractCompressionDecoder
+        implements ClosableCompressionDecoder {
+
+    private static final InternalLogger logger =
+            InternalLoggerFactory.getInstance(AbstractClosableCompressionDecoder.class);
+
+    protected AbstractClosableCompressionDecoder(CompressionFormat format) {
+        super(format);
+    }
+
+    @Override
+    public abstract boolean isClosed();
+
+    @Override
+    protected void handlerRemoved0(ChannelHandlerContext ctx) throws Exception {
+        if (!isClosed()) {
+            logger.warn("ClosableCompressionDecoder is removed from a pipeline, "
+                    + "but the end of the compressed stream has not been reached.");
```

Comment:
> I think it's not really a good idea to raise an exception in handlerRemoved(); it's like failing to close a file.

+1

> Instead, we could make the ClosableCompressionDecoder close the stream automatically when it is removed from the pipeline.

We can not close it, because decoder's state/behavior is determined by inbound compressed stream. And here we have 2 different behaviors:

1.`ClosableCompressionDecoder`s (Bzip2Decoder, Lz4FrameDecoder, JdkZlibDecoder, JZlibDecoder). All of them have similar structure of compressed stream, something like:

| stream-header | block 1 | ... | block n | stream-footer (checksum) |
|----|----|----|----|----|

So, if the user will remove handler during/after decompression of block m (1 <= m < n), and add it again after a few minutes, the stream will be corrupted. Even if the user will add it after a few completed blocks (not during the block) the final checksum in footer will rise an exception.
So, I think that user should not remove handled of `ClosableCompressionDecoder` while it's not already closed. And we may log about it or prevent.

Also notice, that all `ClosableCompressionDecoder`s are not reusable. When it's closed, all new bytes in the stream will be skipped. We may consider a new `reset` method for manual resetting of decoder (and we will use it in `handlerAdded` or doc that user have to call it manually before adding to a pipeline) and constructor parameter (or setter) to allow codec to call `reset` automatically at the end of current compressed stream.

2.`CompressionDecoder`s (LzfDecoder, FastLzFrameDecoder, SnappyFrameDecoder) are all resettable by design and decompresses each new chunk of compressed data again and again. User can easily remove it from pipeline, but decoder may be in illegal state (for example, it's already decoded header but still wait a payload). When the user adds it again to pipeline, stream will be corrupted. In this case, a new `reset` method may be helpful too.